cocos/core/pipeline/render-window.ts

```diff
@@ -0,0 +1,239 @@
+    }
+
+    public static registerCreateFunc (root: Root) {
+        root._createWindowFun = (_root: Root): RenderWindow => new RenderWindow(_root);
+    }
+
+    protected _root: Root;
+    protected _title: string = '';
+    protected _left: number = 0;
+    protected _top: number = 0;
+    protected _width: number = 0;
+    protected _height: number = 0;
+    protected _nativeWidth: number = 0;
+    protected _nativeHeight: number = 0;
+    protected _isOffscreen: boolean = false;
+    protected _renderPass: GFXRenderPass | null = null;
+    protected _colorTexs: GFXTexture[] = [];
+    protected _depthStencilTex: GFXTexture | null = null;
+    protected _framebuffer: GFXFramebuffer | null = null;
+
+    private constructor (root: Root) {
+        this._root = root;
+    }
+
+    public initialize (info: IRenderWindowInfo): boolean {
+
+        if (info.title !== undefined) {
+            this._title = info.title;
+        }
+
+        if (info.left !== undefined) {
+            this._left = info.left;
+        }
+
+        if (info.top !== undefined) {
+            this._top = info.top;
+        }
+
+        if (info.isOffscreen !== undefined) {
+            this._isOffscreen = info.isOffscreen;
+        }
+
+        this._width = info.width;
+        this._height = info.height;
+        this._nativeWidth = this._width;
+        this._nativeHeight = this._height;
+
+        for (let i = 0; i < info.renderPassInfo.colorAttachments.length; i++) {
+            const attachment = info.renderPassInfo.colorAttachments[i];
+            if (attachment.format === GFXFormat.UNKNOWN) {
+                attachment.format = this._root.device.colorFormat;
+            }
+        }
+        if (info.renderPassInfo.depthStencilAttachment) {
+            const attachment = info.renderPassInfo.depthStencilAttachment;
+            if (attachment.format === GFXFormat.UNKNOWN) {
+                attachment.format = this._root.device.depthStencilFormat;
+            }
+        }
+        this._renderPass = this._root.device.createRenderPass(info.renderPassInfo);
+
+        let colorTexturesToCreate = info.renderPassInfo.colorAttachments.length;
+        if (!this._isOffscreen) { colorTexturesToCreate--; } // -1 for swapchain image
+        for (let i = 0; i < colorTexturesToCreate; i++) {
+            const format = info.renderPassInfo.colorAttachments[i].format;
+            const colorTex = this._root.device.createTexture({
+                type: GFXTextureType.TEX2D,
+                usage: GFXTextureUsageBit.COLOR_ATTACHMENT | GFXTextureUsageBit.SAMPLED,
+                format,
+                width: this._width,
+                height: this._height,
+                depth: 1,
+                arrayLayer: 1,
+                mipLevel: 1,
+                flags: GFXTextureFlagBit.NONE,
+            });
+            this._colorTexs.push(colorTex);
+        }
+        if (this._isOffscreen && info.renderPassInfo.depthStencilAttachment) {
```

Comment:
From the check renderPassInfo's depthStencilAttachment could be null, but it's not optional in IGFXRenderPassInfo

Reply:
It is nullable